google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/jpa/test/ParallelInlineIdsIdsOrClauseDeleteHandlerImpl.java

```diff
@@ -0,0 +1,106 @@
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.hql.internal.ast.HqlSqlWalker;
+import org.hibernate.hql.spi.id.inline.IdsClauseBuilder;
+import org.hibernate.hql.spi.id.inline.InlineIdsIdsOrClauseDeleteHandlerImpl;
+import org.hibernate.persister.collection.AbstractCollectionPersister;
+import org.hibernate.type.CollectionType;
+import org.hibernate.type.Type;
+
+/**
+ * Parallel Inline bulk-id delete handler that uses multiple identifier OR clauses.
+ *
+ * @author Vlad Mihalcea
+ * @author Chengyuan Zhao
+ */
+public class ParallelInlineIdsIdsOrClauseDeleteHandlerImpl extends
+    InlineIdsIdsOrClauseDeleteHandlerImpl {
+
+  private final List<String> deletes = new ArrayList<>();
+
+  public ParallelInlineIdsIdsOrClauseDeleteHandlerImpl(
+      SessionFactoryImplementor factory,
+      HqlSqlWalker walker) {
+    super(factory, walker);
+  }
+
+  @Override
+  public String[] getSqlStatements() {
+    return deletes.toArray(new String[deletes.size()]);
+  }
+
+  @Override
+  public int execute(
+      SharedSessionContractImplementor session,
+      QueryParameters queryParameters) {
+
+    IdsClauseBuilder values = prepareInlineStatement(session, queryParameters);
+
+    if (!values.getIds().isEmpty()) {
+      final String idSubselect = values.toStatement();
+
+      for (Type type : getTargetedQueryable().getPropertyTypes()) {
+        if (type.isCollectionType()) {
+          CollectionType cType = (CollectionType) type;
+          AbstractCollectionPersister cPersister = (AbstractCollectionPersister) factory()
+              .getMetamodel().collectionPersister(cType.getRole());
+          if (cPersister.isManyToMany()) {
+            deletes.add(generateDelete(
+                cPersister.getTableName(),
+                cPersister.getKeyColumnNames(),
+                idSubselect,
+                "bulk delete - m2m join table cleanup"
+            ).toStatementString());
+          }
+        }
+      }
+
+      String[] tableNames = getTargetedQueryable().getConstraintOrderedTableNameClosure();
+      String[][] columnNames = getTargetedQueryable().getContraintOrderedTableKeyColumnClosure();
+      for (int i = 0; i < tableNames.length; i++) {
+        // TODO : an optimization here would be to consider cascade deletes and not gen those delete statements;
+        //      the difficulty is the ordering of the tables here vs the cascade attributes on the persisters ->
+        //          the table info gotten here should really be self-contained (i.e., a class representation
+        //          defining all the needed attributes), then we could then get an array of those
+        deletes.add(generateDelete(tableNames[i], columnNames[i], idSubselect, "bulk delete")
+            .toStatementString());
+      }
+
+      // Start performing the deletes
+      deletes.parallelStream().forEach(delete -> {
```

Comment:
I think a vanilla for-loop is better here; it seems like it is only in certain situations where parallelstreams and multithreading would be beneficial. It is uncertain whether this meets the criteria: https://stackoverflow.com/a/20375622/906497

If this saves you from writing this file then I would recommend just omitting it.